### PR TITLE
Intervals

### PR DESCRIPTION
**Pull request by rushk014**
Create functions for interval predictions
- function 1: `mean_prediction, uncertainties = Vset.predict_with_uncertainties(X_test)`
- function 2: `uncertainty, cum_acc = evaluate_uncertainty(uncertainties, mean_predictions, labels)`
- function 3 `sklearn.calibration.calibration_curve(predictions, y_real)`

### CODE REVIEW

**Comment by rushk014** [comment]
`uncertainty, cum_acc = evaluate_uncertainty(uncertainties, mean_predictions, labels)` is not fixed yet but I opened this PR to get some input.

Additionally I was concerned that since `Vset.predict_with_uncertainties(X_test)` is the only function in `Vset` that returns data directly using it might be non-obvious. What do y'all think?

File: notebooks/00_synthetic_classification.ipynb [hunk 1]
```diff
@@ -110,210 +112,33 @@
...
+      "needs_background": "light"
+     },
+     "output_type": "display_data"
     }
    ],
    "source": [
-    "df = dict_to_df(binary_metrics)\n",
-    "df"
+    "mean_preds, uncertainty = modeling_set.predict_with_uncertainties(X_test)\n",
+    "y_test_np = np.array(list(base_dict(y_test).values())).flatten()\n",
```
[the remark this answers is not in the document]
Comment (jpdunc23) [reply notebooks/00_synthetic_classification.ipynb:138]: I think this brings up another issue: getting specific values out of our dicts is annoying, even when the keys are simple like `(y_test,)`. Maybe the solution is to wrap these dictionaries in a class and implement a method like `y_test.get('y_test', 'subsampling_0')` which returns the matching value(s).

File: vflow/convert.py [hunk 2]
```diff
@@ -82,6 +83,38 @@ def dict_to_df(d: dict, param_key=None):
     return df
 
 
+def compute_interval(df: DataFrame, d_label, wrt_label, accum=None):
```
Comment (jpdunc23) [inline vflow/convert.py:86]: This def is repeated on line 118. I think we can probably remove this method since it's functionality is covered by `perturbation_stats`, unless we have other ideas for it (e.g., it could return the interval data in a format that will be easy to plot).

**Review by jpdunc23 — CHANGES_REQUESTED** [review]

File: notebooks/00_synthetic_classification.ipynb (hunk 1 above)
[the remark this answers is not in the document]
Comment (jpdunc23) [reply notebooks/00_synthetic_classification.ipynb:138]: Another way a dict wrapper class could be useful: a lazy Vset's output dict contains values that are `VfuncPromises` and no computation is actually run. If the wrapped output dict is `out`, we could implement a method like `out.eval()` that calls all the `VfuncPromises` in `out` and replaces them with the results.

Let me know if you're interested in working on this (in another branch) and we can start drafting up a design doc.

[the remark this answers is not in the document]
Comment (rushk014) [reply notebooks/00_synthetic_classification.ipynb:138]: Yeah that sounds sweet. I'll fix up the rest of intervals then take a look

**Comment by jpdunc23** [comment]
> `uncertainty, cum_acc = evaluate_uncertainty(uncertainties, mean_predictions, labels)` is not fixed yet but I opened this PR to get some input.

Is the goal is to give output so that we can produce a CAP plot like [this](https://www.geeksforgeeks.org/python-cap-cumulative-accuracy-profile-analysis/)?

File: vflow/vset.py [hunk 3]
```diff
@@ -179,6 +179,13 @@ def predict_proba(self, *args):
             if hasattr(v, 'predict_proba'):
                 pred_dict[k] = v.predict_proba
         return self._apply_func(pred_dict, *args)
+    
+    def predict_with_uncertainties(self, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (rushk014) [reply vflow/vset.py]: Do you mean the computed mean/std should be replicated for each key in the output of `predict_proba` or that there should be a key for each row of the `pred_stats` output?

[the remark this answers is not in the document]
Comment (jpdunc23) [reply vflow/vset.py]: Ya, sorry, I meant the latter.

Actually, it will be better to not use `wrt` and instead add an arg `groups` (also a list) which we pass directly to `perturbation_stats`. We have to be sure the averages are over predictions from matching input data otherwise the output is meaningless, so something like `groups = ['subsampling']` will be common.

[the remark this answers is not in the document]
Comment (jpdunc23) [reply vflow/vset.py]: A few more thoughts:
- It's probably better to return two separate dicts, one for the means and one for the stds, in case we want to go on to use the mean predictions later on.
- This method could just as easily use `predict` instead of `predict_proba`, so maybe we should add a `with_uncertainty=False` arg to `predict` and `predict_proba` and have them call this method (then input to this method would be the predictions dict).
- This is also getting us closer to ensembling: right now we just return an equally-weighted average, by why not do a weighted average based on the training accuracy for each of the groups? Or via some other ensembling scheme?

[the remark this answers is not in the document]
Comment (rushk014) [reply vflow/vset.py]: I like the `with_uncertainty` idea but we would either need to add optional `group_by` and `wrt_col` args to both `predict` and `predict_proba` or force users to call
```python
preds = predict_proba(X_test, with_uncertainty=False)
mean_preds, std_preds = predict_with_uncertainties(preds, group_by=[...], wrt_col='...') 
```
if they want more control over grouping, which seems a little counterintuitive. I just pushed most of the updates you had mentioned (with the exception of the dict wrapper and `with_uncertainty`). Assuming we want to use these dicts in future calls to some Vset I need to recreate the subkeys as `Subkey`s - I can get info for `Subkey.value` and `Subkey.origin` but not `_output_matching` and the `_sep_dicts_id` is lost since the object is new, though I'm not sure thats really important. Do you think that will eventually break?

**Comment by rushk014** [comment]
> > `uncertainty, cum_acc = evaluate_uncertainty(uncertainties, mean_predictions, labels)` is not fixed yet but I opened this PR to get some input.
> 
> Is the goal is to give output so that we can produce a CAP plot like [this](https://www.geeksforgeeks.org/python-cap-cumulative-accuracy-profile-analysis/)?

@jpdunc23 from what I understood I think we wanted something akin to`sklearn.calibration.calibration_curve` sorted by uncertainty, but that seems like a reasonable output. Would we to compute the cumulative sum across binary predictions or the preds probas?

File: vflow/convert.py [hunk 4]
```diff
@@ -84,6 +86,38 @@ def dict_to_df(d: dict, param_key=None):
                 df = df.iloc[:, new_idx]
     return df
 
+def evaluate_uncertainty(mean_preds, std_preds):
```
Comment (jpdunc23) [inline vflow/convert.py:89]: Let's make this work with more general output from `Vset.predict_with_uncertainties`, i.e. when `group_by` is used, by returning `uncertainty` and `cum_acc` as a list with one entry for each matched entry in `mean_preds` + `std_preds`. Then we can, for example, plot two models separately:
```
u, c = evaluate_uncertainty(mean_dict, std_dict)
plt.plot(u[0], c[0]) # DT
plt.plot(u[1], c[1]) # LR
```

File: vflow/vset.py [hunk 5]
```diff
@@ -179,6 +179,42 @@ def predict_proba(self, *args):
             if hasattr(v, 'predict_proba'):
                 pred_dict[k] = v.predict_proba
         return self._apply_func(pred_dict, *args)
+    
+    def predict_with_uncertainties(self, *args, group_by: list=None, proba: bool=True):
```
Comment (jpdunc23) [inline vflow/vset.py:183]: Let's make `proba: bool=False` the default.

**Review by jpdunc23 — CHANGES_REQUESTED** [review]

File: vflow/vset.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jpdunc23) [reply vflow/vset.py]: @rushk014 ya I think this could become an issue, though I can't come up with anything off of the top of my head right now. But some changes I have in mind to Subkeys and dict wrapping may solve the problem. For now, can you try create a reproducible example and include it in a new issue?

**Review by jpdunc23 — APPROVED** [review]
